codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -135,53 +135,114 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
+                        frameHeader.release();
+                        frameHeader = null;
+                    }
+                } else {
+                    remainingData -= maxFrameSize;
+                    // Write the header.
+                    ByteBuf lastFrame;
+                    if (frameHeader == null) {
+                        lastFrame = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
+                        writeFrameHeaderInternal(lastFrame, maxFrameSize, DATA, flags, streamId);
+                    } else {
+                        lastFrame = frameHeader.slice();
+                        frameHeader = null;
+                    }
+                    ctx.write(lastFrame, promiseAggregator.newPromise());
+
+                    // Write the payload.
+                    lastFrame = data.readSlice(maxFrameSize);
+                    data = null;
+                    ctx.write(lastFrame, promiseAggregator.newPromise());
                 }
-                if (needToReleaseData) {
+
+                do {
+                    int frameDataBytes = min(remainingData, maxFrameSize);
+                    int framePaddingBytes = min(padding, max(0, (maxFrameSize - 1) - frameDataBytes));
+
+                    // Decrement the remaining counters.
+                    padding -= framePaddingBytes;
+                    remainingData -= frameDataBytes;
+
+                    // Write the header.
+                    ByteBuf frameHeader2 = ctx.alloc().buffer(DATA_FRAME_HEADER_LENGTH);
+                    flags.endOfStream(endStream && remainingData == 0 && padding == 0);
+                    flags.paddingPresent(framePaddingBytes > 0);
+                    writeFrameHeaderInternal(frameHeader2, framePaddingBytes + frameDataBytes, DATA, flags, streamId);
+                    writePaddingLength(frameHeader2, framePaddingBytes);
+                    ctx.write(frameHeader2, promiseAggregator.newPromise());
+
+                    // Write the payload.
+                    if (frameDataBytes != 0) {
+                        if (remainingData == 0) {
+                            ByteBuf lastFrame = data.readSlice(frameDataBytes);
+                            data = null;
+                            ctx.write(lastFrame, promiseAggregator.newPromise());
+                        } else {
+                            ctx.write(data.readRetainedSlice(frameDataBytes), promiseAggregator.newPromise());
+                        }
+                    }
+                    // Write the frame padding.
+                    if (paddingBytes(framePaddingBytes) > 0) {
+                        ctx.write(ZERO_BUFFER.slice(0, paddingBytes(framePaddingBytes)),
+                                  promiseAggregator.newPromise());
+                    }
+                } while (remainingData != 0 || padding != 0);
+            }
+        } catch (Throwable cause) {
+            if (frameHeader != null) {
+                frameHeader.release();
```

Comment:
we also have the other release() call in a `try {} finally { }` block. Should we do the same here ?

Reply:
The `frameHeader` is created internally so I don't think it is necessary. The `try/finally` for `data` is to cover the case where it has already been released when passed to this method. I will add comments.

Reply:
I see... @Scottmitch pull in once you added the comment.